## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
# Imported PANDA library.
import pandas as pd

# Imported the data table.
df = pd.read_csv('Resources/client_dataset.csv')

# Displayed the data table.
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data

# Printed the names of the columns in the DataFrame.
print(df.columns)


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [3]:
# Use the describe function to gather some basic statistics

# Displayed the summary of general statistics.
df.describe()


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.

# Created dictionary by mapping original column names to new column names.
column_name_change = {
    'first': 'First Name',
    'last': 'Last Name',
    'job': 'Occupation',
    'phone': 'Phone Number',
    'email': 'Email Address',
    'client_id': 'Client ID',
    'order_id': 'Order ID',
    'order_date': 'Order Date',
    'order_week': 'Order Week',
    'order_year': 'Order Year',
    'item_id': 'Item ID',
    'category': 'Category',
    'subcategory': 'Sub-Category',
    'unit_price': 'Unit Price',
    'unit_cost': 'Unit Cost',
    'unit_weight': 'Unit Weight',
    'qty': 'Quantity',
    'line_number': 'Line Number',
}

# Renamed the columns in the DataFrame for better readability.
df.rename(columns=column_name_change, inplace=True)

# Displayed the DataFrame to ensure changed column names.
df.head()


First Name Last Name             Occupation        Phone Number  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                         Email Address  Client ID  Order ID  Order Date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   Order Week  Order Year         Item ID     Category        Sub-Category  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   Unit Price  Unit Cost  Unit Weight  Quantity  Line Number  
0     1096.80     762.71         7.50       105            1  
1       24.95      15.09         1.49        21            0  
2       13.52       7.86         1.68        39            6  
3       36.42      24.85         1.23        29            3  
4      195.10     108.17        46.43        20            1

In [8]:
# How many rows of data are there? (Continuation of above for research purpose.)

# Fetched the number of rows in the DataFrame.
num = df.shape[0]

# Printed the number of rows.
print("Number of Rows:", num)


Number of Rows: 54639


In [9]:
# How many clients are named Christina? (Continuation of above for research purpose.)
name_counts = df['First Name'].value_counts()

# Get the count of occurrences for the name 'Christina.'
christina_count = name_counts.get('Christina', 0)

# Printed the number of people named Christina.
print("Number of people named Christina:", christina_count)


Number of people named Christina: 210


In [10]:
# What three item categories had the most entries?

# Counted the occurrences of each entry in the 'Category' column in the DataFrame
cat_counts = df['Category'].value_counts()

# Displayed the top 3 entries.
cat_counts.head(3)

Category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [11]:
# For the category with the most entries,
# which subcategory had the most entries?

# Identified the category with the highest number of entries.
cat_refrence = cat_counts.head(1).index[0]

# Filtered the DataFrame for entries with the identified 'Category.'
refrence_df = df[df['Category'] == cat_refrence]

# Counted the occurrences of each 'Sub-Category' within the identified category.
sub_count = refrence_df['Sub-Category'].value_counts()

# Found the 'Sub-Category' with the highest count.
sub_max = sub_count.idxmax()

# Printed the highest consumable sub-category and the number of purchases.
print("Highest consumable:", sub_max)
print("Number of purchases:", sub_count[sub_max])


Highest consumable: bathroom supplies
Number of purchases: 6424


In [12]:
# Which five clients had the most entries in the data?

# Counted the occurrences of each client ID in the DataFrame.
client_id_count = df['Client ID'].value_counts()

# Displayed the top 5 client IDs with the highest number of entries.
client_id_count.head(5)


Client ID
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [13]:
# Stored the Client IDs of top 5 clients in a list.
client_id_count.head(5).index.tolist()


[33615, 66037, 46820, 38378, 24741]

In [14]:
# How many total units (the qty column) did the
# client with the most entries order?

# Fetched the 'Client ID' with the highest number of entries.
refrence_id = client_id_count.head(1).index[0]

# Filtered the DataFrame for aggregate entries with the identified 'Client ID.'
id_df = df[df['Client ID'] == refrence_id]

# Calculated the total quantity purchased for the selected 'Client_ID.'
qty_value = id_df['Quantity'].sum()

# Printed the 'Client ID' and the corresponding total quantity purchased.
print(f"Client ID: {refrence_id}, entry total:", qty_value)


Client ID: 33615, entry total: 64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [15]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty

# Created a new column 'Line Subtotal' by multiplying 'Unit Price' and 'Quantity.'
df['Line Subtotal'] = df['Unit Price'] * df['Quantity']

# Selected relevant columns for display.
select_columns = ['Unit Price', 'Quantity', 'Line Subtotal']

# Displayed the DataFrame with the selected columns.
df[select_columns].head()


Unit Price  Quantity  Line Subtotal
0     1096.80       105      115164.00
1       24.95        21         523.95
2       13.52        39         527.28
3       36.42        29        1056.18
4      195.10        20        3902.00

In [16]:
# I created a new column, 'Total Weight' that references: 'Unit Weight' and 'Quantity'
df['Total Weight'] = df['Unit Weight'] * df['Quantity']

# Displayed the first few rows to ensure 'Total Weight' was created properly.   
df.head()

First Name Last Name             Occupation        Phone Number  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                         Email Address  Client ID  Order ID  Order Date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   Order Week  Order Year         Item ID     Category        Sub-Category  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   Unit Price  Unit Cost  Unit Weight  Quantity  Line Number  Line Subtotal  \
0     1096.80     762.71         7.50       105            1      115164.00   
1       24.95      15.09         1.49        21            0         523.95   
2       13.52       7.86         1.68        39            6         527.28   
3       36.42      24.85         1.23        29            3        1056.18   
4      195.10     108.17        46.43        20            1        3902.00   

   Total Weight  
0        787.50  
1         31.29  
2         65.52  
3         35.67  
4        928.60

In [17]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.

# Defined a function to calculate the shipping price based on 'Total Weight.'
def calculate_shipping_price(total_weight):
    if total_weight > 50:
        return 7 * total_weight
    else:
        return 10 * total_weight

# Applied the shipping price calculation function to the 'Total Weight' column.
df['Shipping Price'] = df['Total Weight'].apply(calculate_shipping_price)

# Selected relevant columns for display.
select_columns_b = ['Unit Price', 'Unit Weight', 'Quantity', 'Total Weight', 'Shipping Price']

# Displayed the first few rows with the selected columns.    
df[select_columns_b].head()


Unit Price  Unit Weight  Quantity  Total Weight  Shipping Price
0     1096.80         7.50       105        787.50         5512.50
1       24.95         1.49        21         31.29          312.90
2       13.52         1.68        39         65.52          458.64
3       36.42         1.23        29         35.67          356.70
4      195.10        46.43        20        928.60         6500.20

In [18]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%

# Created a column 'Total Price' for each line adding 'Line Subtotal'to 
# 'Shipping Price', and added a sales tax of 9.25%.
df['Total Price'] = ((df['Line Subtotal'] + df['Shipping Price']) * 1.0925).round(2)

# Selected relevant columns for display.
select_columns_c = ['Line Subtotal', 'Shipping Price', 'Total Price']

# Displayed the first few rows with the selected columns.
df[select_columns_c].head()

Line Subtotal  Shipping Price  Total Price
0      115164.00         5512.50    131839.08
1         523.95          312.90       914.26
2         527.28          458.64      1077.12
3        1056.18          356.70      1543.57
4        3902.00         6500.20     11364.40

In [19]:
# Create a column for the cost 
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).

# Created a column 'Total Cost' for each line multiplying 'Unit Cost' with 'Quantity', 
# and adding 'Shipping Price." Then rounded output data to 2 decimals.
df['Total Cost'] = (df['Unit Cost'] * df['Quantity'] + df['Shipping Price']).round(2)

# Selected relevant columns for display.
select_columns_d = ['Unit Cost', 'Quantity', 'Shipping Price', 'Total Price', 'Total Cost']

# Displayed the first few rows with the selected columns.
df[select_columns_d].head()

Unit Cost  Quantity  Shipping Price  Total Price  Total Cost
0     762.71       105         5512.50    131839.08    85597.05
1      15.09        21          312.90       914.26      629.79
2       7.86        39          458.64      1077.12      765.18
3      24.85        29          356.70      1543.57     1077.35
4     108.17        20         6500.20     11364.40     8663.60

In [21]:
# Create a column for the profit of
# each line using line cost and line price

# Created a column, 'Gross Profit' for each line using 'Total Price' to subtract 'Total Cost.'
df['Gross Profit'] = (df['Total Price'] - df['Total Cost']).round(2)

# Selected relevant columns for display.
select_columns_e = ['Total Price', 'Total Cost', 'Gross Profit']

# Displayed the first 5 rows of the selected columns.
df[select_columns_e].head()

Total Price  Total Cost  Gross Profit
0    131839.08    85597.05      46242.03
1       914.26      629.79        284.47
2      1077.12      765.18        311.94
3      1543.57     1077.35        466.22
4     11364.40     8663.60       2700.80

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [23]:
# Check your work using the totals above.

# Listed 'Order ID' values.
id_search = [2742071, 2173913, 6128929]

# Calculated the sum for each 'Order ID' value.
for order_id in id_search:
    order_id_rows = df[df['Order ID'] == order_id].index
    tp_sum = df.iloc[order_id_rows]['Total Price'].sum()
    usd_tp_sum = '{:,.2f}'.format(tp_sum)

    # Displayed results.
    print(f"'Total Price' for 'Order ID' {order_id} is: ${usd_tp_sum}")


'Total Price' for 'Order ID' 2742071 is: $152,811.89
'Total Price' for 'Order ID' 2173913 is: $162,388.71
'Total Price' for 'Order ID' 6128929 is: $923,441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [24]:
# How much did each of the top 5 clients by quantity spend? 
# Check your work from Part 1 for client ids.

# Referenced 'client_id_count' from Part 1.
top_5 = client_id_count.head(5).index

# Filtered the DataFrame for the top 5 clients.
top_5_df = df[df['Client ID'].isin(top_5)]

# Calculated the total spending for the top 5 clients.
top_5_total_spend = top_5_df.groupby('Client ID')['Total Price'].sum()

# Formated the total spending as a dollar amount.
usd_total_spend = top_5_total_spend.map('${:,.2f}'.format)

# Displayed the total spending for the top 5 clients.
print(usd_total_spend)


Client ID
24741    $82,268,892.02
33615     $8,377,308.52
38378    $12,906,550.87
46820     $9,743,794.36
66037    $10,259,514.79
Name: Total Price, dtype: object


In [53]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.

# Created a DataFrame with totals for the top 5 clients.
sum_a_df = pd.DataFrame({
    'Total Quantity Purchased': top_5_df.groupby('Client ID')['Quantity'].sum(),
    'Total Shipping Price': top_5_df.groupby('Client ID')['Shipping Price'].sum(),
    'Total Price': top_5_df.groupby('Client ID')['Total Price'].sum(),
    'Total Profit': top_5_df.groupby('Client ID')['Gross Profit'].sum()
}).round(2)

# Displayed the DataFrame with totals for the top 5 clients.
sum_a_df.head()


Total Quantity Purchased  Total Shipping Price  Total Price  \
Client ID                                                                
24741                        239862            5126448.37  82268892.02   
33615                         64313            1828984.89   8377308.52   
38378                         73667            3429455.40  12906550.87   
46820                         75768            1601448.84   9743794.36   
66037                         43018            1395151.85  10259514.79   

           Total Profit  
Client ID                
24741       36579992.31  
33615        2201994.61  
38378        3271829.89  
46820        2736602.72  
66037        3255031.81

In [25]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.

# Previous column names are adequately formatted.
# Formatted previous DataFrame to display in millions $USD to 3 decimal places.
mm_sum_a_df = pd.DataFrame({
    'Total Quantity Purchased': top_5_df.groupby('Client ID')['Quantity'].sum(),
    'Total Shipping Price (MM)': (top_5_df.groupby('Client ID')['Shipping Price'].sum() / 1_000_000).map('${:,.3f}'.format),
    'Total Price (MM)': (top_5_df.groupby('Client ID')['Total Price'].sum() / 1_000_000).round(2).map('${:,.3f}'.format),
    'Total Profit (MM)': (top_5_df.groupby('Client ID')['Gross Profit'].sum() / 1_000_000).round(2).map('${:,.3f}'.format)
})

# Displayed the updated DataFrame
mm_sum_a_df.head()


Total Quantity Purchased Total Shipping Price (MM)  \
Client ID                                                       
24741                        239862                    $5.126   
33615                         64313                    $1.829   
38378                         73667                    $3.429   
46820                         75768                    $1.601   
66037                         43018                    $1.395   

          Total Price (MM) Total Profit (MM)  
Client ID                                     
24741              $82.270           $36.580  
33615               $8.380            $2.200  
38378              $12.910            $3.270  
46820               $9.740            $2.740  
66037              $10.260            $3.260

In [26]:
# Sort the updated data by "Total Profit" form highest to lowest.

# Ammended above data in descending order.
sort_mm_sum_a_df = mm_sum_a_df.sort_values(by='Total Profit (MM)', ascending=False)

# Displayed the updated DataFrame
sort_mm_sum_a_df.head()

Total Quantity Purchased Total Shipping Price (MM)  \
Client ID                                                       
24741                        239862                    $5.126   
38378                         73667                    $3.429   
66037                         43018                    $1.395   
46820                         75768                    $1.601   
33615                         64313                    $1.829   

          Total Price (MM) Total Profit (MM)  
Client ID                                     
24741              $82.270           $36.580  
38378              $12.910            $3.270  
66037              $10.260            $3.260  
46820               $9.740            $2.740  
33615               $8.380            $2.200